### ISSUE TITLE
Adding native mixture training

### ISSUE THREAD

**Issue by shermansiu** [huggingface/trl#1050]
Mixture training, where the pre-training objective (e.g. next token prediction) is mixed in with the alignment objective can help improve the overall performance of the model.

According to the [InstructGPT paper](https://arxiv.org/abs/2203.02155) and the [DeepSpeed Chat Readme](https://github.com/microsoft/DeepSpeed/blob/master/blogs/deepspeed-chat/README.md), it can prevent regression on benchmarks like SQuADv2 and DROP.

**Comment by lvwerra** [comment]
Adding a CLM loss to the current implementation should be fairly straightforward to add. It would require passing the pretraining data sample plus some parameters to control the ratio and data sampling. Would you be interested in working on a PR?

**Comment by 1485840691-eng** [comment]
If no one is working on this, I would like to try it.

DeepSpeed has implemented CLM loss and we might follow [it](https://github.com/microsoft/DeepSpeedExamples/blob/8e4cdd88574a1b3ce2f9f73c30b7ad3434ff1866/applications/DeepSpeed-Chat/training/step3_rlhf_finetuning/main.py#L556).

To prepare pre-trained data samples, it creates a separate data loader and zip with the prompt data loader, so every batch will fetch both prompt and pre-trained data sample. Moreover, it creates a separate [MiniDataset ](https://github.com/microsoft/DeepSpeedExamples/blob/8e4cdd88574a1b3ce2f9f73c30b7ad3434ff1866/applications/DeepSpeed-Chat/dschat/utils/data/data_utils.py#L485)to support mini-batch training.

In our case, we could as well load pretrained data and create separate data loader as DeepSpeed. We shall modify the step() function to accept PretrainData including input_ids, attention_mask, labels.

    @PPODecorators.empty_device_cache()
    def step(
        self,
        queries: List[torch.LongTensor],
        responses: List[torch.LongTensor],
        scores: List[torch.FloatTensor],
        response_masks: Optional[List[torch.LongTensor]] = None,
        **pretrained_datas: Optional[List[PretrainData]] = None,**
    ):
@dataclass
class PretrainData:
    input_ids: Optional[List[torch.LongTensor]] = None,
    attention_mask: Optional[List[torch.LongTensor]] = None,
    labels: Optional[List[torch.LongTensor]] = None,

We might as well inject the unsupervised loss after [train_minibatch](https://github.com/huggingface/trl/blob/20428c48ba651b2bc11b2a73e9eb9568b1af3f96/trl/trainer/ppo_trainer.py#L795).

While in loss implementation, we might borrow logic from iterative_sft_trainer. 

Will add an extra parameter for CLM loss coefficient.

What do you think? Any suggestions and feedback is appreciated.
